src/disco/archiver/fd_archiver_rocksdb.c

```diff
@@ -0,0 +1,283 @@
+  fd_alloc_t *           alloc;
+  fd_valloc_t            valloc;
+};
+typedef struct fd_archiver_rocksdb_tile_ctx fd_archiver_rocksdb_tile_ctx_t;
+
+FD_FN_PURE static inline ulong
+loose_footprint( fd_topo_tile_t const * tile FD_PARAM_UNUSED ) {
+  return 2UL * FD_SHMEM_GIGANTIC_PAGE_SZ;
+}
+
+static void
+rocksdb_inspect( fd_archiver_rocksdb_tile_ctx_t * ctx,
+                 fd_stem_context_t *              stem ) {
+  ulong start_slot = 0;
+  ulong end_slot   = 0;
+  ulong shred_cnt  = 0;
+  do {
+    if( FD_UNLIKELY( fd_rocksdb_root_iter_next( &ctx->rocksdb_root_iter, &ctx->rocksdb_slot_meta, ctx->valloc ) ) ) break;
+    if( FD_UNLIKELY( fd_rocksdb_get_meta( &ctx->rocksdb, ctx->rocksdb_slot_meta.slot, &ctx->rocksdb_slot_meta, ctx->valloc ) ) ) break;
+    ulong slot = ctx->rocksdb_slot_meta.slot;
+    ulong start_idx = 0;
+    ulong end_idx = ctx->rocksdb_slot_meta.received;
+
+    rocksdb_iterator_t * iter = rocksdb_create_iterator_cf(ctx->rocksdb.db, ctx->rocksdb.ro, ctx->rocksdb.cf_handles[FD_ROCKSDB_CFIDX_DATA_SHRED]);
+
+    char k[16];
+    *((ulong *) &k[0]) = fd_ulong_bswap(slot);
+    *((ulong *) &k[8]) = fd_ulong_bswap(start_idx);
+
+    rocksdb_iter_seek(iter, (const char *) k, sizeof(k));
+
+    for (ulong i = start_idx; i < end_idx; i++) {
+      ulong cur_slot, index;
+      uchar valid = rocksdb_iter_valid(iter);
+
+      if (valid) {
+        size_t klen = 0;
+        const char* key = rocksdb_iter_key(iter, &klen); // There is no need to free key
+            if (klen != 16)  // invalid key
+              FD_LOG_ERR(( "rocksdb has invalid key length" ));
+            cur_slot = fd_ulong_bswap(*((ulong *) &key[0]));
+            index = fd_ulong_bswap(*((ulong *) &key[8]));
+      }
+
+      if (!valid || cur_slot != slot)
+        FD_LOG_ERR(("missing shreds for slot %lu, valid=%u", slot, valid));
+
+      if (index != i)
+        FD_LOG_ERR(("missing shred %lu at index %lu for slot %lu", i, index, slot));
+
+      size_t dlen = 0;
+      // Data was first copied from disk into memory to make it available to this API
+      const unsigned char *data = (const unsigned char *) rocksdb_iter_value(iter, &dlen);
+      if (data == NULL)
+        FD_LOG_ERR(("failed to read shred %lu/%lu", slot, i));
+
+      // This just correctly selects from inside the data pointer to the
+      // actual data without a memory copy
+      fd_shred_t const * shred = fd_shred_parse( data, (ulong) dlen );
+      if( start_slot==0 ) start_slot = shred->slot;
+      end_slot = shred->slot;
+      shred_cnt++;
+
+      fd_blockstore_shred_insert( ctx->blockstore, shred );
+      if( !!(shred->data.flags & FD_SHRED_DATA_FLAG_DATA_COMPLETE) ) {
+        int slot_complete = !!(shred->data.flags & FD_SHRED_DATA_FLAG_SLOT_COMPLETE);
+        /* Notify the replay tile after inserting a FEC set */
+        FD_LOG_INFO(( "%lu:[%u, %u] notifies replay", shred->slot, shred->fec_set_idx, shred->idx ));
+        uint  cnt   = shred->idx+1-shred->fec_set_idx;
```

Comment:
this isn't correct -- you're not watermarking with `cnt` correctly. replay expects an entry batch, which can span more than one FEC set. this is assuming cnt is the FEC cnt of the current shred